neuralmagic/benchmarks/scripts/backend_request_func.py

```diff
@@ -98,58 +105,99 @@ async def async_request_tgi(
-            "top_p": 1.0,
-            "max_tokens": request_func_input.output_len,
-            "ignore_eos": True,
-            "stream": True,
-        }
-        output = RequestFuncOutput()
-        output.prompt_len = request_func_input.prompt_len
-
-        ttft = 0
-        st = time.perf_counter()
-        try:
-            async with session.post(url=api_url, json=payload) as response:
-                if response.status == 200:
-                    data = None
-                    async for part_data in response.content.iter_any():
-                        if ttft == 0:
-                            ttft = time.perf_counter() - st
-                            output.ttft = ttft
-                        data = part_data
-                    output.latency = time.perf_counter() - st
-
-                    # When streaming, '\0' is appended to the end.
-                    body = trim_suffix(data.decode('utf-8'), "\0")
-                    output.generated_text = json.loads(
-                        body)["text"][0][len(request_func_input.prompt):]
-
-                    output.success = True
-
-                else:
-                    output.success = False
-        except (aiohttp.ClientOSError, aiohttp.ServerDisconnectedError):
-            output.success = False
-
-        if pbar:
-            pbar.update(1)
-        return output
+class AsyncRequestVLLM:
+
+    @staticmethod
+    def stream_server_outputs():
+        """
+        This function is queried to set the `stream` option of the 
+        server JSON payload.
+        """
+        return True
+
+    @staticmethod
+    def decode_server_response(server_response: bytes, prompt_len: int) -> str:
+        """
+        Decodes the server response and returns the text generated by
+        the server in response to the prompt.
+        """
+
+        def try_json_decode(s: str) -> dict:
+            try:
+                return json.loads(s)
+            except Exception as _:
+                return None
+
+        # When streaming, '\0' is appended to the end.
+        assert (AsyncRequestVLLM.stream_server_outputs())
+        body: str = trim_suffix(server_response.decode('utf-8'), "\0")
+
+        # Most times we only have one JSON in the body.
+        decoded_json = try_json_decode(body)
+        if decoded_json is not None:
+            return decoded_json["text"][0][prompt_len:]
+
+        # Some times body contains more than one JSON.
+        # These JSONs essentially contain the generated text and the
+        # last of the JSONs has the entire generated text.
+        json_starts = [m.start() for m in re.finditer('{\"text\":', body)]
+        for json_start in reversed(json_starts):
+            decoded_json = try_json_decode(body[json_start:])
+            if decoded_json is not None:
+                return decoded_json["text"][0][prompt_len:]
+
+        raise ValueError(f"Cannot decode json body \n {body}")
+
+    @staticmethod
+    async def async_request_vllm(
```

Comment:
no functional change this function except storing the response directly in the RequestFuncOutput